### PR TITLE
chain: go borsh library to encode struct to bytes

### LINKED ISSUE #1198: Go borsh library to encode struct to bytes
Create `Encoder` to encode each field of a struct.

### PR DESCRIPTION
**Pull request by tansawit**
Fixes: #1198

### CODE REVIEW

File: chain/borsh/decode.go [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+		offset: 0,
+	}
+}
+
+// Finished returns a bool on whether the decoding has finished
+func (decoder *BorshDecoder) Finished() bool {
+	return decoder.offset == uint32(len(decoder.data))
+}
+
+// DecodeU8 deodes the input bytes and returns the corresponding
```
Comment (taobun) [inline chain/borsh/decode.go]: deodes

File: chain/borsh/decode.go [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+		return 0, errors.New("Borsh: out of range")
+	}
+	val := binary.LittleEndian.Uint64(decoder.data[decoder.offset : decoder.offset+8])
+	decoder.offset += 8
+	return val, nil
+}
+
+// DecodeSigned8 deodes the input bytes and returns the corresponding signed
+// `int8` value and any errors
+func (decoder *BorshDecoder) DecodeSigned8() (int8, error) {
```
Comment (taobun) [inline chain/borsh/decode.go]: DecodeI8

File: chain/borsh/decode.go [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+	unsigned, _ := decoder.DecodeU8()
```
Comment (taobun) [inline chain/borsh/decode.go]: Why ignore error?

File: chain/borsh/decode.go [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+	unsigned, _ := decoder.DecodeU64()
+	return int64(unsigned), nil
+}
+
+// DecodeBytes deodes the input bytes and returns the corresponding
+// `[]bytes` slice and any errors
+func (decoder *BorshDecoder) DecodeBytes() ([]byte, error) {
+	length, err := decoder.DecodeU32()
+	if err != nil {
+		return []byte{}, err
```
Comment (taobun) [inline chain/borsh/decode.go]: return nil, err

**Review by taobun — CHANGES_REQUESTED** [review]

File: chain/borsh/decode_test.go [hunk 5]
```diff
@@ -0,0 +1,313 @@
...
+
+	if decodeFinished {
+		t.Errorf(`Incorrect result for Finished bool: expected "%t" got "%t"`, false, decodeFinished)
+	}
+}
+
+func TestFinishedExpectTrue(t *testing.T) {
+	data, _ := hex.DecodeString("10")
+	decoder := BorshDecoder{data, 1}
+	fmt.Println(len(decoder.data))
```
Comment (taobun) [inline chain/borsh/decode_test.go]: Remove this

File: chain/borsh/decode_test.go [hunk 6]
```diff
@@ -0,0 +1,313 @@
...
+	return NewBorshDecoder(data)
+}
+
+func TestFinishedExpectFalse(t *testing.T) {
+	data, _ := hex.DecodeString("03000000425443")
+	decoder := NewBorshDecoder(data)
+	decodeFinished := decoder.Finished()
+
+	if decodeFinished {
+		t.Errorf(`Incorrect result for Finished bool: expected "%t" got "%t"`, false, decodeFinished)
```
Comment (taobun) [inline chain/borsh/decode_test.go]: Our code base use [require](https://godoc.org/github.com/stretchr/testify/require) for testing

File: chain/borsh/decode_test.go [hunk 7]
```diff
@@ -0,0 +1,313 @@
...
+	fmt.Println(len(decoder.data))
+	decodeFinished := decoder.Finished()
+
+	if !decodeFinished {
+		t.Errorf(`Incorrect result for Finished bool: expected "%t" got "%t"`, true, decodeFinished)
+	}
+}
+
+func TestDecodeU8(t *testing.T) {
+	var expectedU8 uint8 = 112
```
Comment (taobun) [inline chain/borsh/decode_test.go]: expectedU8 := uint8(112)

File: chain/borsh/decode_test.go [hunk 8]
```diff
@@ -0,0 +1,313 @@
...
+	decoder := getDecoder("70")
+	resultU8, err := decoder.DecodeU8()
+
+	if err != nil {
```
Comment (taobun) [inline chain/borsh/decode_test.go]: If we use `require`, it will just require.NotNil(t, err)

File: chain/borsh/encode.go [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+func NewBorshEncoder() BorshEncoder {
+	return BorshEncoder{
+		data: []byte{},
+	}
+}
+
+// EncodeU8 takes a `uint8` variable and encodes it into a byte array
+func (encoder *BorshEncoder) EncodeU8(value uint8) []byte {
+	bytes := []byte{value}
+	encoder.data = append(encoder.data, bytes...)
```
Comment (taobun) [inline chain/borsh/encode.go]: We can write `encoder.data = append(encoder.data, value)`

**Review by taobun — CHANGES_REQUESTED** [review]

File: chain/borsh/decode_test.go [hunk 10]
```diff
@@ -0,0 +1,211 @@
...
+	data, _ := hex.DecodeString(inputString)
+	return NewBorshDecoder(data)
+}
+
+func TestFinishedExpectFalse(t *testing.T) {
+	data, _ := hex.DecodeString("03000000425443")
+	decoder := NewBorshDecoder(data)
+	decodeFinished := decoder.Finished()
+
+	require.Equal(t, decodeFinished, false, `Incorrect result for Finished boolean: expected "false"`)
```
Comment (taobun) [inline chain/borsh/decode_test.go]: require.False

File: chain/borsh/encode.go [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+
+// EncodeU32 takes a `uint32` variable and encodes it into a byte array
+func (encoder *BorshEncoder) EncodeU32(value uint32) {
+	bytes := make([]byte, 4)
+	binary.LittleEndian.PutUint32(bytes, value)
+	encoder.data = append(encoder.data, bytes...)
+}
+
+// EncodeU64 takes a `uint64` variable and encodes it into a byte array
+func (encoder *BorshEncoder) EncodeU64(value uint64) {
```
Comment (sorawit) [inline chain/borsh/encode.go]: an

**Review by taobun — CHANGES_REQUESTED** [review]

**Review by sorawit — CHANGES_REQUESTED** [review]

**Review by sorawit — APPROVED** [review]
Looks good

**Review by taobun — APPROVED** [review]

**Pull request merged**